### ISSUE TITLE
Significant Packet Delay with quic-go on iOS Due to Goroutine Handling

### ISSUE THREAD

**Issue by minchopm** [quic-go/quic-go#4557]
I'm experiencing significant network traffic delay when using the quic-go library for quic connections on iOS. The issue appears to stem from quic-go's internal use of goroutines. This problem seems to be closely related to a known issue with thread handling in Go, as described in [this Go issue](https://github.com/golang/go/issues/65889). The same behaviour is observed when using webtransport-go (version 0.8.0) for packet transmission.

Steps to Reproduce:

Set up a quic connection using the quic-go library with the Dial method.
Generate 1000-byte packets and send them every 20 milliseconds.
Compile and run the Go code on iOS using the following build command:
bash
Copy code
GOOS=ios GOARCH=arm64 CGO_ENABLED=1 CC=$(go env GOROOT)/misc/ios/clangwrap.sh go build -ldflags -w -v -tags ios -o "IOS.a" -buildmode=c-archive
Observe significant packet delay in network traffic.
Expected Behaviour:
Network traffic should be handled correctly without packet delay, even with quic-go's internal use of goroutines.

Actual Behavior:
When using quic-go on iOS, there is significant packet delay in network traffic, likely due to the interaction between quic-go's goroutines and the Go runtime's thread handling on iOS.

Additional Information:

The issue seems to be related to how the Go runtime handles thread locking and goroutine switching, particularly in the context of iOS. The details are extensively discussed in [this Go issue](https://github.com/golang/go/issues/65889).
While we are using the Dial method to open a quic connection, a similar behaviour can be observed when wrapping a UDP connection in a goroutine, resulting in network traffic delay. However, without using goroutines, there is no issue with network traffic.
The same behaviour is observed when sending packets via [webtransport-go](https://github.com/quic-go/webtransport-go) (version 0.8.0).
The problem seems to be exacerbated by quic-go's heavy internal use of goroutines, interacting poorly with the Go runtime's thread scheduling on iOS.
Environment:

quic-go version: 0.44.0
Go version: 1.22
Operating System: iOS
Build Command:

bash
Copy code
GOOS=ios GOARCH=arm64 CGO_ENABLED=1 CC=$(go env GOROOT)/misc/ios/clangwrap.sh go build -ldflags -w -v -tags ios -o "IOS.a" -buildmode=c-archive
Relevant Links:

Related Go issue: [golang/go#65889](https://github.com/golang/go/issues/65889)
Request for Assistance:
I'm seeking assistance or a workaround for this issue. If this is a known problem with a potential fix or if there's a recommended approach to manage the internal goroutine usage in quic-go on iOS, please provide guidance.

Thank you for your attention to this matter.

**Comment by marten-seemann** [comment]
> The problem seems to be exacerbated by quic-go's heavy internal use of goroutines, interacting poorly with the Go runtime's thread scheduling on iOS.

Can you elaborate what you mean by heavy use? We only have one Go routine per connection, plus one for the send queue (https://github.com/quic-go/quic-go/blob/master/connection.go#L515).

You could try driving the send queue from the connection's Go routine, this should be a very simple change.

**Comment by minchopm** [comment]
Thank you for your response. I apologize for any confusion caused by my previous explanation. Upon further investigation, I realized that the issue isn't related to the heavy use of goroutines within quic-go.

To clarify, the problem appears to be with the use of even a single goroutine causing significant packet delay on iOS. Here are the details of my findings:

Testing with a Regular UDP Connection:

I set up a basic UDP connection and sent 1000-byte packets every 20 milliseconds.
When running this setup without using any goroutines, the network traffic is handled correctly with no significant delays.
However, when I wrapped the UDP connection inside a single goroutine, I observed significant packet delays.
Implications for quic-go:

Given that quic-go uses a single goroutine per connection plus one for the send queue, this behavior likely affects quic-go as well, since it involves goroutine handling.

Build Environment:

I used the following build command for iOS:

```
GOOS=ios GOARCH=arm64 CGO_ENABLED=1 CC=$(go env GOROOT)/misc/ios/clangwrap.sh go build -ldflags -w -v -tags ios -o "IOS.a" -buildmode=c-archive
```

It seems the interaction between the Go runtime's thread scheduling and goroutines on iOS is the root cause of the delay, as extensively discussed in [this Go issue](https://github.com/golang/go/issues/65889).

I appreciate your suggestion to drive the send queue from the connection's goroutine and will look into implementing this change to see if it mitigates the issue.

Thank you for your attention and assistance with this matter.
